### PR TITLE
Deduplicate literal -> constant lowering

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

File: src/librustc_mir/hair/pattern/mod.rs [hunk 1]
```diff
@@ -1296,19 +1294,32 @@ pub fn compare_const_vals<'a, 'tcx>(
...
-                          tcx: TyCtxt<'a, 'tcx, 'tcx>,
-                          ty: Ty<'tcx>,
-                          neg: bool)
-                          -> Result<&'tcx ty::Const<'tcx>, LitToConstError> {
+pub fn lit_to_const<'a, 'gcx, 'tcx>(
+    lit: &'tcx ast::LitKind,
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    ty: Ty<'tcx>,
+    neg: bool,
+) -> Result<&'tcx ty::Const<'tcx>, LitToConstError> {
```
Comment (eddyb) [inline src/librustc_mir/hair/pattern/mod.rs]: Can this be moved out of `hair::pattern`? Somewhere else in `hair` would be fine (cc @nikomatsakis).

Comment (oli-obk) [reply src/librustc_mir/hair/pattern/mod.rs]: I created a custom file, there's likely other things we'll want to add in the future.

**Comment by eddyb** [comment]
Ehm I think I wanted to say this (but it's not that big of a deal):

r? @nikomatsakis

**Pull request merged**